frontend/src/pages/shipments/ShipmentOffers.tsx

```diff
@@ -0,0 +1,131 @@
+import cx from 'classnames'
+import { FunctionComponent, useMemo } from 'react'
+import { Column, useSortBy, useTable } from 'react-table'
+import Badge from '../../components/Badge'
+import ButtonLink from '../../components/ButtonLink'
+import TableHeader from '../../components/table/TableHeader'
+import {
+  OffersForShipmentQuery,
+  useAllGroupsMinimalQuery,
+  useOffersForShipmentQuery,
+} from '../../types/api-types'
+import { offerCreateRoute } from '../../utils/routes'
+
+interface Props {
+  shipmentId: number
+}
+
+const ShipmentOffers: FunctionComponent<Props> = ({ shipmentId }) => {
+  // Get the list of offers for this shipment
+  const { data } = useOffersForShipmentQuery({
+    variables: { shipmentId },
+  })
+
+  // Get the list of groups
+  const { data: groups } = useAllGroupsMinimalQuery()
+
+  const offers = useMemo(() => data?.listOffers || [], [data])
+
+  const COLUMNS = useMemo(() => {
+    if (groups?.listGroups && data?.listOffers) {
+      const columns: Column<OffersForShipmentQuery['listOffers'][0]>[] = [
```

Comment:
question (non-blocking): what's going on with the type here?

Reply:
Welcome to my hell! First, the point of the whole `Column<...>` type here is to get a nice API when defining table columns with `react-table`:
```ts
{
  Header: 'Pallets',
  // Here, TS knows that `row` is has a `pallets` property that's an array
  accessor: (row) => row.pallets.length,
}
```

So why does it look so gross? Because the codegen doesn't generate a _named type_ from the schema. To break it down:
```ts
Column<OffersForShipmentQuery['listOffers'][0]>
// ^            ^                  ^        ^
// 1            2                  3        4
```
1. `Column` comes from `react-table` and allows us to define type-safe... columns that are passed to the hook
2. `OfferForShipmentQuery` is the shape of the data that's returned from the `useOfferForShipmentQuery` hook:

```ts
type OffersForShipmentQuery = { __typename?: 'Query' } & {
  listOffers: Array<
    { __typename?: 'Offer' } & Pick<
      Offer,
      'id' | 'status' | 'sendingGroupId' | 'photoUris'
    > & { pallets: Array<{ __typename?: 'Pallet' } & Pick<Pallet, 'id'>> }
  >
}

// Extrapolated version:
type OffersForShipmentQuery = {
  __typename?: 'Query'
  listOffers: Array<{
    __typename?: 'Offer'
    id: number
    status: OfferStatus
    sendingGroupId: number
    photoUris: string[]
    pallets: Array<{
      __typename?: 'Pallet'
      id: number
    }>
  }
}
```

3. In the `OffersForShipmentQuery` type above, the `listOffers` property is what we care about: that's the array of offers we wanna display.
4. Finally, the `listOffers[0]` is what tells TS that we're looking for the shape of a single array element. It doesn't matter what the value of the index is.

**Can we do better?** Probably! Maybe there's a way to generate named types from `graphql-codegen`.

I've also tried relying on the `Offer` type but that's not compatible with the shape of the items in `OffersForShipmentQuery.listOffers`. We would have to pick the properties we care about and it would become even wordier.

Reply:
Oh wow, thank you for the very clear explanation. Yeah it sounds like there's not a great option here. Nice work maintaining type safety and not just bypassing the checks!